distributed/shuffle/_buffer.py

```diff
@@ -0,0 +1,231 @@
+                            self.sizes[part_id] -= s
+                        except IndexError:
+                            break
+                        finally:
+                            if not self.shards[part_id]:
+                                del self.shards[part_id]
+                                assert not self.sizes[part_id]
+                                del self.sizes[part_id]
+                else:
+                    shards = self.shards.pop(part_id)
+                    size = self.sizes.pop(part_id)
+                self._shards_available.notify_all()
+            await self.process(part_id, shards, size)
+
+    async def put(self, data: dict[str, list[ShardType]]) -> None:
+        """
+        Writes many objects into the local buffers, blocks until ready for more
+
+        Parameters
+        ----------
+        data: dict
+            A dictionary mapping destinations to lists of objects that should
+            be written to that destination
+        """
+
+        if self._exception:
+            raise self._exception
+        if self._closed or self._done:
+            raise RuntimeError(f"Trying to put data in closed {self}.")
+
+        if not data:
+            return
+
+        shards = None
+        size = 0
+
+        sizes = {}
+        for id_, shards in data.items():
+            size = sum(map(len, shards))
+            sizes[id_] = size
+        total_batch_size = sum(sizes.values())
+        self.bytes_memory += total_batch_size
+        self.bytes_total += total_batch_size
+
+        if self.memory_limiter:
+            self.memory_limiter.acquire(total_batch_size)
+        async with self._shards_available:
+            for id_, shards in data.items():
+                self.shards[id_].extend(shards)
+                self.sizes[id_] += sizes[id_]
+            self._shards_available.notify()
+        if self.memory_limiter:
+            await self.memory_limiter.wait_for_available()
+        del data, shards
+        assert size
+
+    def raise_on_exception(self) -> None:
+        if self._exception:
+            raise self._exception
+
+    async def flush(self) -> None:
+        """Wait until all writes are finished"""
+        async with self._flush_lock:
+            self._closed = True
+            async with self._shards_available:
+                await self._shards_available.wait_for(
+                    lambda: not self.shards or self._exception or self._done
+                )
+                self._done = True
+                self._shards_available.notify_all()
+
+            await asyncio.gather(*self._tasks)
+            if not self._exception:
+                assert not self.bytes_memory, (type(self), self.bytes_memory)
+
+    async def close(self) -> None:
+        await self.flush()
+        if not self._exception:
+            assert not self.bytes_memory, (type(self), self.bytes_memory)
+        for t in self._tasks:
+            t.cancel()
+        self._closed = True
```

Comment:
`flush` also sets `_closed` and `_done`. Should we clean this up so that `flush` is not responsible for setting these, just for checking that the buffer is indeed closed?

Reply:
In a perfect world, flush would not necessarily close the buffer at all. however, untangling this proved to be a bit more difficult than I thought it would be. I'll keep this as is, for now. This PR is already big enough as it is.